### DISCUSSION TITLE
Evennia Coding Challenges

### DISCUSSION THREAD

**Discussion by kovitikus** [evennia/evennia#2295] [General Discussion]
### ATTENTION!
**Only post a new message directly to this main thread if it is a challenge.** To respond to a challenge, there are separate reply boxes below each main post that will create a sub-thread. Place any questions, suggestions, or ideas regarding that specific challenge within the sub-thread.

# Introduction

The purpose of this thread is to gather ideas for coding challenges that the community can tackle, specifically those new to the engine.

Challenges are unlike tutorials. They aren't step-by-step and laid out, holding your hand. They are somewhere in-between a tutorial and the cold dark empty void where you're left to your own devices. If you are new to Evennia, these challenges will help teach you how to navigate the Evennia API and documentation. They should also give you the confidence to solve problems on your own. These two skills may be the most critical things you can master!

**Do not cheat!** By cheating, I mean looking at another person's solution before you attempt to carve your own path. You will only hurt yourself! The point of these challenges is to think, and copying another person's solution requires no thinking.

Ask questions! Flood this thread with questions, they are not considered cheating. But, try your best to solve your own problem before asking for help. I will fully admit that I wouldn't have gotten anywhere with Evennia if I didn't have helpful folk answering my questions, but I also had to find the right balance between being handed an answer and seeking out my own.

There are a lot of resources already written for Evennia and my hope is that these challenges will help point you in the right direction when it comes to utilizing these resources.

I would also like to ask that you try to keep your questions and answers contained within this thread. It's not wrong to ask questions in live chat, such as Discord or IRC, but those questions and answers will get lost to history. It's best that we keep as much information contained herein as possible, to avoid having the same questions asked multiple times.

# Challenge Layout
A challenge should be laid out in the following manner, as to best facilitate conveying the idea and maximizing resource references. Be certain to include any tidbits that might not be explained in other sections of Evennia; little tips or quirks when managing your code's logic that is critical to its function.

* A brief summary of what the challenge is, including a desired outcome.
* An estimated rating of the difficulty of the challenge, ranging from 1 (easy) to 5 (difficult).
* The challenge should be broken down into smaller tasks, such as method examples. Use `##` for main titles and `####` for method sections.
* Include mentions of Evennia functions and methods that would be useful for accomplishing specified tasks. It helps to provide the full Evennia path to the method you are mentioning.
* A conclusion section where you can elaborate on additional things that can be done with the challenge.

**Please post your solutions in the sub-thread of the challenge post; even if you are the author of the challenge.** This will help keep the solution slightly detached from the main challenge post and make it less likely for the reader to ruin their challenge by clicking on the solution prematurely!

The posted solution should be "hidden". Anybody taking the coding challenge should not check the solution until they feel they've completed the challenge to the best of their ability. Coders can then check their results against the posted solution to compare their outcome to others.

Use this [HTML code](https://github.com/evennia/evennia/discussions/2267#discussioncomment-200461) to post a solution.
```
<details>
<summary>
Click here for the solution!
</summary>

Here is the solution I developed for the challenge.

(Triple backtick block code markdown should go here.)
func MyCode(arguments):
    if code:
        return solution
(Triple backtick block code markdown should go here.)

Explanations for why you did certain things is welcomed.

Be sure to post all of your code related to the challenge 
and mark where it is placed in the project, to track organization.
</details>
```

<details>
<summary>
Click here for the solution!
</summary>
Here is the solution I developed for the challenge.

```python
func MyCode(arguments):
    if code:
        return solution
```
Be sure to post all of your code related to the challenge 
and mark where it is placed in the project, to track organization.
</details>

**Comment by kovitikus** [comment]
# Torch Challenge - 1/5 Difficulty

#### Challenge Summary
Make a torch typeclass that has a fuel attribute. Add to it an ignite method which causes a message to echo to the room and a ticker to begin eating away at the fuel until the torch dies. Each time the ticker executes the specified method, check that there is remaining fuel and echo a message to the room. Once dead, destroy the torch.

Create a new command that calls upon the torch's ignite method, getting the process started.

## Torch Object
In `typeclasses/objects.py`, create a new class for your torch and inherit Object or DefaultObject.

#### def at_object_creation(self):
Within this method add your fuel attribute, as well as a rate for how much fuel will be consumed per tick.

#### def ignite(self, igniter):
Create an `ignite` method and be sure to specify the character that is lighting it as an argument, so that you can refer to them.

Send a message to the igniting character that they've lit the torch. You should also send a message to all of the room's contents, using the `msg_contents()` method and be sure to exclude the igniter!

Add a tickerhandler and set the seconds to the frequency which you wish the torch to burn at. This is also the frequency that will echo the messages to the room, letting characters in the room know that the torch is still burning. In the tickerhandler, also add the method you wish to call per tick. I used `self.on_burn_tick`, because this method also lives on the torch and self is a reference to the torch object. You should also set the tickerhandler to be persistent, as you will likely want it to continue to burn even after the server is reloaded.

#### def on_burn_tick(self):
Start out by getting the attributes for the total fuel value and fuel burn rate from the torch and assigning those to variables.

Now it's time to do some conditional checks. We need to know if the fuel has enough left to burn another tick, so the first thing we want to check for is if the fuel is less than or equal zero. If that's the case, we should call upon `self.death`, which is a method I'll get to shortly.

Now that we know the torch has enough fuel to burn, since it passed our death check, we can echo a message to the room via `msg_contents()` method and let characters know that the torch is still burning. The flavor of the text is up to you!

We also need to set the torch's fuel attribute to `-=` our fuel burn rate, so that there's less fuel or none at all the next time we tick!

> **TIP:** You should know that it's possible to rearrange the logic of this ticker. You can check to see if the fuel has less than or equal to the fuel_burn_rate before any other logic and have one final flicker of flame before it dies. It's just a matter of how you want to handle the behavior. The way I originally wrote it, the torch burns with 0 fuel left until the next tick. Strive to write better logic than I did!

#### def death(self):
Now that the torch is dead you should remove the `tickerhandler`.

Next, you need to echo a death message for the torch to the room, letting players know that it's finished.

Finally, use `self.delete()` to destroy the torch object!

## Ignite Command
You have the logic written for what happens with the torch, but you need a way to initiate this logic. This can be done via a command.

Create a new command in your `commands/command.py` module.

#### def parse(self):
The parsing method of a command is used to determine the validity of the arguments passed to the command and to exit the command prematurely if a condition is not met.

Commands have a built-in property on them that holds the object (usually a character) that called the command. This is `self.caller`. Self refers to the command object and caller is the property on that object.

Commands also have another property that tracks the arguments passed to it, this is `self.args`.

In the parse method, we need to check to see if there are any arguments at all. If there are no arguments stored in `self.args`, send an error message to the caller and use `raise InterruptCommand` to exit the command early. `raise InterruptCommand` must also be imported before using it, so ensure that you've done so. Its path is `evennia.InterruptCommand`.

`InterruptCommand` is an exception that allows us to gracefully exit the parsing method, preventing the `def func(self):` method from being executed and altogether exiting the command without any further code being evaluated. It is only used in `parse` and not in `func`!

If `self.args` has some value, we now need to evaluate the arguments. You should know that args is a string and that it includes anything typed directly after the command, including whitespace. 

Generally you will have extra whitespace on the leftside of an argument, so you want to clean that up with `.strip()`. This strips whitespace from both the beginning and trailing ends of the string, but not from the middle. You can specify `.lstrip()` and `.rstrip()` instead if you wish to only strip whitespace from one end or the other.

Now that you've got a clean string, it's time to search for the object that best represents the string passed to the command. In this case, we are looking to search for a torch. 

Out in the wilds of your code, you may have to use a utility like `evennia.utils.search.search_object`, but when you have an object nearby, such as a the `caller`, you can use the handy little method from `evennia.objects.objects.DefaultObject.search`. Utilize this search to find an object that matches the arguments. After the search, you should check to see if your object variable that you stored the search results on `is not None`. If it is `None`, send an error message to the caller and `raise InterruptCommand`.

Now that you have your torch object stored in a variable, I highly suggest checking it to make sure that it is the proper typeclass. This will prevent unexpected behavior by having characters trying to ignite things you don't want ignited!

There are several ways to check if an object is the proper type of object you are looking for; tags are one example of something I use quite often. In this case, we know that we have a custom typeclass of Torch and so we can use that to check.

For this check, we will utilize `evennia.typeclasses.models.TypedObject.is_typeclass`. This built-in method doesn't need to be imported and can be accessed directly on the object we are checking.

If the object isn't the Torch typeclass, send an error message to the caller and raise InterruptCommand. Else, if it is the Torch typeclass, assign the object to `self.variable`, such as `self.obj` to access it in the `def func()` method of the command object.

#### def func(self):
We have all of our parsing out of the way and we are certain that we have an object stored and that it is of the proper typeclass. Now it's time to execute our command's logic. In this case, the logic is simply to call the ignite method of the torch object.

## Additional Thoughts
Viola! There you have it. A functioning torch that can be ignited, echos for a bit to the room, and then dies.

There's a few things that can be done to beef up the torch from here. You should really try to make this challenge your own and wow the rest of us with your fantastic torch/lantern/flashlight, whatever it is you want to do!

For me, I wanted to track the location of the torch. Is it in the hands of a player or is it lying on the floor of a room? If a player is holding it, they should see a different output from the torch than the other folk in the room. Try tracking the torch's `.location` and implementing that into your logic to display different messages.

Something else you might want to do, that I have yet to implement, is have your room object track lighting and get your torch to interact with the lighting attribute of the room. See if you can make the room lighter or darker. There are a multitude of ways to handle this and the lighting attribute may not even need to be placed on the room itself. See if you can take it even further and change how players can interact with the room, such as the look command, based on the lighting.

**Reply by kovitikus** [reply]
This is where you post anything regarding the torch challenge, including solutions. Please keep all comments regarding the torch challenge within this sub-thread and only post new challenges to the main thread.

**Reply by kovitikus** [reply]
<details>
<summary>
Here's my solution for the torch challenge!
</summary>

My `objects.py` Torch class.

Originally, and in the long term, I only want my torches to burn out after 10 minutes of use.

For quicker test results, I made it burn out in 2 minutes instead; ticking for 25/100 fuel every 30 seconds.

```python
class Torch(Lighting):
    def at_object_creation(self):
        self.attributes.add('fuel', 100)
        # Torches burn at the rate of 10 fuel per minute, or 5 fuel every 30 seconds.

        self.tags.add('stackable')
        self.tags.add('inventory', category='stack')

    def ignite(self, lighter):
        room, held_by = self.find_location()
        
        lighter.msg(f"You light {self.name}, igniting it.")
        room.msg_contents(f"{lighter.get_display_name(self)} lights {self.name}, igniting it.", exclude=lighter)

        tickerhandler.add(30, self.on_burn_tick, persistent=True)

    def on_burn_tick(self):
        fuel = self.attributes.get('fuel')
        room, held_by = self.find_location()
        burn_rate = 25

        if fuel <= 0:
            self.death(room, held_by)
            return
        
        if inherits_from(held_by, "typeclasses.characters.Character"):
            held_by.msg(f"Your {self.name} flickers.")
            room.msg_contents(f"{held_by.name}'s {self.name} flickers.", exclude=held_by)
        else:
            room.msg_contents(f"{self.name} flickers.")
        self.db.fuel -= burn_rate
            
    def death(self, room, held_by):
        tickerhandler.remove(30, self.on_burn_tick, persistent=True)
        if inherits_from(held_by, "typeclasses.characters.Character"):
            held_by.msg(f"Your {self.name} torch dies out and it collapses into a tiny pile of ash.")
            room.msg_contents(f"{held_by.name}'s {self.name} dies out and it collapses into a tiny pile of ash.", exclude=held_by)
        else:
            room.msg_contents(f"{self.name} dies out and it collapses into a tiny pile of ash.")
        self.delete()

    def find_location(self):
        #Check to see if the torch is in the hands (inventory) of player
        # or if the torch is in a a room.
        held_by = self.location
        room = None

        if inherits_from(held_by, 'typeclasses.characters.Character'):
            room = held_by.location
        elif inherits_from(held_by, 'typeclasses.rooms.Room'):
            room = held_by
        return room, held_by
```

My `command.py` Ignite command.

```python
class CmdLight(Command):
    key = 'light'

    def parse(self):
        caller = self.caller
        if not self.args:
            caller.msg("Light what?")
            raise InterruptCommand
        else:
            args = self.args.strip()
            obj = caller.search(args, quiet=True)
            obj = obj[0]
            if obj:
                if not inherits_from(obj, 'typeclasses.objects.Lighting'):
                    caller.msg(f"{obj.get_display_name(caller)} cannot be lit!")
                    raise InterruptCommand
                else:
                    self.obj = obj
            else:
                caller.msg(f"{args} not found!")
                raise InterruptCommand

    def func(self):
        caller = self.caller
        obj = self.obj

        obj.ignite(caller)
```

</details>

**Reply by Griatch** [reply]
<details>
<summary>
My take on the torch challenge.
</summary>

This variation includes a `light/douse` double-functioning command, as well as uses Attributes to configure the torch so it can be customized by a builder in-game (including setting custom messages on it). To use, the `CmdManageTorch` needs to be added to the default Character cmdset. I put typeclass and command in the same file since they are so tightly coupled.

```python
# in a new file mygame/typeclasses/torch.py

from random import choice, random
from typeclasses.objects import Object
from evennia import TICKER_HANDLER
from evennia import default_cmds, InterruptCommand


class Torch(Object):
    """
    A simple torch that burns out after a certain time. The reason for using a ticker here 
    is because we want the torch to give off random messages and to burn out 
    on its own without any input from the wielder.

    """

    def at_object_creation(self):
        """
        Setup Attributes to control and tweak the torch by bulders in-game.
        """

        self.db.burning = False
        self.db.desc = "A simple torch, its tip wrapped in oiled rags."
        self.db.fuel = 60
        self.db.burn_rate_per_minute = 1
        self.db.burn_messages = ["The {key} flickers.", "The {key}'s flame flickers.'"]
        self.db.end_messages = ["The {key} burns out.", "The {key} flickers and dies."]

    def _msg(self, msg):
        """
        Send message to carrier and carrier's location.
        """
        carrier = self.location
        if carrier:
            # this would be the carrier's inventory
            carrier.msg(msg)
            if carrier.location:
                # the room
                carrier.location.msg_contents(
                    msg + f" (carried by {carrier.key})",
                    exclude=carrier)

    def check_burn(self):
        """
        Called by TickerHandler every minute to decide if the torch still
        burns. Also gives some atmospheric messages.

        """
        self.db.fuel -= self.db.burn_rate_per_minute
        if self.db.fuel <= 0 and self.douse():
            # torch is already doused at this point
            msg = choice(self.db.end_messages).format(key=self.key)
            self._msg(msg)
            # delete a burned-out torch
            self.delete()
        elif self.db.burning:
            # show a random message 10% of the time
            if random.random() < 0.1:
                self._msg(choice(self.db.burn_messages).format(key=self.key))

    def ignite(self):
        """
        Ignite the torch.

        Returns:
            bool: True if this action ignited the torch, False if
                it couldn't be lit.

        """
        if self.db.fuel:
            self.db.burning = True

            TICKER_HANDLER.add(
                interval=60,
                callback=self.check_burn,
                idstring=self.id,
                persistent=True
            )

        return self.db.burning

    def douse(self):
        """
        Douse the torch, if it was burning.

        Returns:
            bool: True if the dousing action was succesful, False
                if it failed (because it wasn't burning).

        """
        if self.db.burning:
            self.db.burning = False

            TICKER_HANDLER.remove(
                interval=60,
                callback=self.check_burn,
                idstring=self.id,
                persistent=True
            )

            return True
        return False


class CmdManageTorch(default_cmds.MuxCommand):
    """
    Manage a carried torch.

        light <carried torch>
        douse <carried torch>

    """

    key = "light"
    aliases = ["ignite", "douse"]

    def parse(self):
        super().parse()
        if not self.args:
            self.caller.msg("Usage: light/douse <carried torch>")
            raise InterruptCommand
        self.torch = self.caller.search(self.args, location=self.caller)
        if not self.torch:
            raise InterruptCommand
        self.light_action = self.cmdname in ['light', 'ignite']

    def func(self):

        if self.light_action:
            if self.torch.ignite():
                self.msg(f"You light {self.torch.key}.")
            else:
                self.msg("This torch has burned out.")
        else:
            if self.torch.douse():
                self.msg(f"You douse {self.torch.key}.")
            else:
                self.msg("This torch is not burning...")

```

</details>

**Comment by alephate** [comment]
Here's a coding challenge, though I likely won't write it in as detailed a manner as Kovitikus did.

### "Make A Cat You Can Pet" Challenge - 1/10 Difficulty 

**Challenge Summary**
The objective of this challenge is to complete it as simply as possible, using basic Evennia functionality! There needs to be a cat object, and a command to pet it. You must be able to pet the cat. Every time someone pets the cat, the cat's affection for them increases, and the cat responds differently based on how much it loves the person who is petting it.

**Challenge Process**
Use any way you want to complete the challenge requirements. At the core, it consists of three things: a cat object, an attribute stored on either the cat or the character that signifies affection, and a pet command.

**Reply by alephate** [reply]
<details>
<summary>Here's my take on the cat challenge! 
</summary>

First, while logged in, I used Evennia admin CmdCreate and then used `py` to edit the cat object made:
```
create/drop a cat
py cat = self.search("cat");cat.tags.add("cat");cat.db.affection = {}
```

Then, made this command and added it to my CharacterCmdSet:

```

class CmdPetCat(Command):
    """
    Pet something!

    Usage: pet thing

    """

    key = "pet"

    def func(self):
        if not self.args:
            self.caller.msg("What do you want to pet?")
            return
        thing_to_pet = self.caller.search(self.args)
        if not thing_to_pet:
            # There is nothing to pet by that name! The error message is already sent.
            return
        # Send a petting emote to the room.
        self.caller.location.msg_contents(f"{self.caller.key} pets {thing_to_pet.key}.")
        if not thing_to_pet.tags.get("cat"):
            # If the thing to pet is a cat, it will be tagged so. Otherwise uh oh, it's not a cat.
            # Maybe it's another PC!
            return
        # Cat affection will be stored as a dictionary, in the following format: {person.key: integer}
        if self.caller.key not in thing_to_pet.db.affection.keys():
            # This person has never petted this cat before! We will add them to the cat's affection dictionary with an
            # initial value of 1
            current_affection = thing_to_pet.db.affection[self.caller.key] = 1
        else:
            # We will increase the affection that the cat has for this person
            current_affection = thing_to_pet.db.affection[self.caller.key] + 1
            thing_to_pet.db.affection[self.caller.key] = current_affection
        # Depending on the cat's affection for the person, we will send an emote.
        if current_affection > 100:
            self.caller.location.msg_contents(f"{thing_to_pet.key} purrs at {self.caller.key}.")
        elif current_affection > 50:
            self.caller.location.msg_contents(f"{thing_to_pet.key} rubs against {self.caller.key}'s hand.")
        elif current_affection > 20:
            self.caller.location.msg_contents(f"{thing_to_pet.key} arches stiffly.")
        else:
            self.caller.location.msg_contents(f"{thing_to_pet.key} hisses.")
      ```

</details>

**Reply by InspectorCaracal** [reply]
<details>
<summary>
PETTING THE CAT
</summary>

Here we go!

First, the `pet` command. I made it as generic as possible so that you can pet whatever you want. Put it in a new file or whatever and then add it to your Character cmdset

```python
class CmdPet(Command):
    """
    Pet someone or something.
    
    Usage:
        pet <target>
    """
    key = "pet"
    
    def func(self):
        # find the target
        target = self.caller.search(args)
        if not target:
            return
        
        # send the pet emote message to the room regardless! if someone wants to pet a chair, who are we to stop them
        self.caller.location.msg_contents("$You() $conj(pet) $you(target)", mapping={'target': target}, from_obj=self.caller)

        # check if it's got special petting responses
        if hasattr(target, "at_pet"):
            target.at_pet(self.caller)
```

Next up, we have an `AffectionateObject`, well, object. It's just like a normal object, but it has that `at_pet` method. I put it in `typeclasses/affection.py`

```python
from .objects import Object

class AffectionateObject(Object):
    """An object that does things when you pet it!"""

    def at_pet(self, petter, **kwargs):
        """React to being pet"""
        # this doesn't work if there's no petter
        if not petter:
            return

        # do we have any affection yet?
        if not affection := self.db.affection:
            # initialize the attribute to the necessary dictionary
            # we can initialize it with our first affection point!
            self.db.affection = { petter: 1 }
            val = 1
            # re-retrieve the attribute to our variable
            affection = self.db.affection

        else:
            # retrieve our affection for petter
            val = affection.get(petter, 0) + 1
            # increment it by one
            affection[petter] = val
        
        # we have an affection value - now we have to retrieve the right response
        # this is some fancy shenanigans for integer-keyed dictionaries to retrieve the maximum valid item
        responses = sorted(self.attributes.get('pet_responses', {}).items(), key=lambda x: x[0])
        responses = [ res for i, res in responses if i <= val ]
        # if we have valid responses, grab the last one and use it
        if responses:
            response = responses[-1]
            self.location.msg_contents(response, from_obj=self, mapping={'petter': petter})
            
        # otherwise, do nothing
```

Last but definitely not least, we'll define our Pettable Cat prototype! Stick this into `world/prototypes.py`

```python
PETTABLE_CAT = {
    'typeclass': 'typeclasses.affection.AffectionateObject',
    'key': 'cat',
    # here's the fun part!
    'pet_responses': {
            0: "$You() $conj(hiss) at $you(petter).",
            2: "$You() $conj(tolerate) $you(petter)'s affection.",
            5: "$You() $conj(purr) contentedly at $you(petter).",
        }
}
```

Reload everything, then spawn a cat in game and try petting it! I didn't actually test it so here's hoping I didn't mess up!
</details>
